### ISSUE TITLE
Yarn --productions tries to resolve devDeps as well

### ISSUE THREAD

**Issue by Diokuz** [yarnpkg/yarn#3630]
**Do you want to request a *feature* or report a *bug*?**
This is a bug, or unexpected behaviour.

**What is the current behavior?**
1. Have a `package.json` which have `devDependencies` and one of them are private
2. Run `yarn --production`
3. Yarn tries to resolve all packages and fails, because have no access to private packages

**What is the expected behavior?**
Yarn shoud only resolve production packages when `--production` flag exposed.

**Please mention your node.js, yarn and operating system version.**
```
node -v
v8.1.0
yarn --version
0.24.6
```

**Comment by rally25rs** [comment]
This happens because Yarn needs to consider the peerDependencies so that it can produce a deterministic build between prod and dev builds (packages will be hoisted to the same locations).

**Comment by Diokuz** [comment]
Not sure I got it... lets consider two situations:

1. We have `package.json` with deps and devDeps, and we run `yarn --production`
2. We remove all devDeps from that `package.json` and run `yarn`

Will there be the difference in `node_modules`? If yes, why?

**Comment by Diokuz** [comment]
Actually, I think this is related to my another problem https://github.com/yarnpkg/yarn/issues/1379

**Comment by rally25rs** [comment]
Possibly. Consider something like:

```
dependencies {
  A@1.0.0
}
devDependencies {
  B@2.0.0
}
```

And `A` also has a dependency:

```
dependencies {
  B@1.0.0
}
```

-----

With devDependencies, the dependency tree is:

```
A@1.0.0
|- B@1.0.0
B@2.0.0
```

so A@1 can not longer be hoisted to top level, resulting in:

```
node_modules/
|- A/ (1.0.0)
|  |- node_modules/
|     |- B/ (1.0.0)
|- B/ (2.0.0)
```

-----

Without devDependencies, the dependency tree is:

```
A@1.0.0
|- B@1.0.0
```

But `B` can be hoisted to the top level, so the result is:

```
node_modules/
|- A/ (1.0.0)
|- B/ (1.0.0)
```

(This is non-deterministic because B@1 could end up in 2 different paths; either in `node_modules/B` or `node_modules/A/node_modules/B` depending on the type of install)

-----

What Yarn wants to do for a deterministic build is to make sure that B@1.0.0 is installed to the same location every time, so a `--production` build should notice that `B@2.0.0` will be hoisted to top level _if_ the devDependencies were there, so the resulting install should really be:

```
node_modules/
|- A/ (1.0.0)
   |- node_modules/
      |- B/ (1.0.0)
```

-----

I hope that makes sense :) Basically all the devDependencies are built into the dependency tree when figuring out where to hoist packages to.

**Comment by rally25rs** [comment]
@Diokuz I haven't tried it myself, but you might see if NPM5 has the same issue. If I understand the NPM5 lockfile correctly, I think it saves the resolved / hoisted position for each package in their lockfile (Yarn does not), so it might not need to query the metadata for your private repos. As much as I hate to steer anyone toward NPM, this might be a case where NPM5 works better. It's worth a test anyway!

You might also be able to solve this using the Yarn offline mirror, since the metadata and package would be saved there and shouldn't have to query the actual server. However that might have other security concerns for you, since a copy of your private repo code would be in that mirror directory.

**Comment by Diokuz** [comment]
@rally25rs thank you so much for explanation!

So, I think I get my answer. The only question now: will that be done in near future? I mean, saving positions in lock file.

**Comment by arcanis** [comment]
It's not only a question of position. Let's say you have this package.json:

```
{
    "dependencies": {
        "a": "^1.0.0"
    },
    "devDependencies": {
        "b": "^1.0.0"
    }
}
```

And that the following are true:

  - `a@1.0.0` and `a@1.1.0` both exist
  - `b@1.0.0` exists and depends on `a@1.0.0`

If we don't resolve devDependencies, then in prod we'll have the following:

```
/node_modules/
/node_modules/a (@1.1.0)
```

And in dev, because of the deduplication optimization, we'll have the following:

```
/node_modules/
/node_modules/a (@1.0.0)
/node_modules/b (@1.0.0)
```

Note that we end up using a different version of `a` in prod than in dev - that might not be good.

**Comment by arcanis** [comment]
Now that I think about this, maybe in practice it could work, since the lockfile would still pin the version - at least assuming that this lockfile has been generated in dev mode.

Still, I'm not convinced this is an issue worth fixing right now. I feel like this is something that could probably be better solved through your build environment.

**Comment by Diokuz** [comment]
Well, we have solved our problem via `NPM_TOKEN` env var. But if _npm_ is better here, why dont you want to make _yarn_ even better in near (or far) future?)

Anyway, thanks for answers)

**Comment by kentor** [comment]
@arcanis We don't control the build environment. For example I am using Netlify which does not have access to private dependencies through github (not npm private registry, so I can't do something like `NPM_TOKEN`). I thought it would be ok to move the private deps to `devDependencies` but yarn in production mode still tries to resolve it. The lock file should be enough to determine the positions of the packages.

**Comment by saikojosh** [comment]
I love Yarn but this has repeatedly been raised as an issue over the last year and is always closed. Surely this is a common use case to have access to different packages in production and development?

My own situation is that I'm using a private Git dependency in development and in CI/production I am using a private module hosted on NPM. I don't want to give my CI system or production systems access to GitHub and so Yarn should not need access to that dependency. If NPM can figure out a solution that works, so can Yarn.

For now I am going to have to switch back to NPM.

**Comment by Zephir77167** [comment]
Hey, bumping into this issue at well.

```
yarn --version
0.27.5
```

How to reproduce:

- Create an empty folder, `cd` into it
- Create a `package.json`:
```
{
  "devDependencies": {
    "doesnt-exist": "file:./nope"
  }
}
```
- Run: `yarn --production`

`npm i --production` does work.

**Comment by BYK** [comment]
Have any of you tried using `--prefer-offline` or `--offline`? That should work IMO.

**Comment by Zephir77167** [comment]
Just tried, doesn't seem to work.

**Comment by olingern** [comment]
@BYK Opened a PR for this fix: #4210

**Comment by thedaniel** [comment]
That PR has been closed without merging, any idea when a new one might make it onto the schedule?

**Comment by olingern** [comment]
The issue is slightly complex. I believe the solution could be attaching metadata of sorts to the lockfile, but I defer to @arcanis @BYK

**Comment by BYK** [comment]
The thing is, we still need those `package.json` files to have a consistent image of the resolved file tree, even if those dependencies won't be installed.

An alternative would be to use the offline mirror feature that would include the dev dependency packages or, their stripped-down versions which only contains the package.json files.

@olingern's suggestion also sounds good but I don't know what the implications for that are. @kaylieEB, @arcanis, @bestander - any thoughts?

**Comment by rally25rs** [comment]
I would imagine we could add a "node_modules path where the packages is hoisted to" in the metadata. I think that would eliminate the need to re-query the metadata to build the hoisted dep tree. To me the bigger issue is compatibility. Maybe this becomes something for a yarn v2 roadmap?

**Comment by bestander** [comment]
I think there are a lot of ways to go around this issue: linking, workspaces, fake offline mirror + tweaking yarn.lock.
And this issue is IMHO not common enough to justify some significant changes to package resolver which is the core of Yarn.

At this stage a better solution is to document the easiest way to go around this limitation and send a PR to our blog https://github.com/yarnpkg/website/tree/master/_posts

**Comment by Zephir77167** [comment]
I have found several people with this problem, which doesn't occur on NPM. Sometimes you have a local dependency which you don't want to include in production, and you have to use NPM on the production env (docker or such), which can lead to problems, caused by the differences between Yarn and NPM. Could you please ELI5 why this works on NPM but not on Yarn? Thanks!

**Comment by bestander** [comment]
Yarn is designed to produce the same results on all computers, it verifies that all dependencies are present and fold/deduplicate the same way.
E.g. 

```
A -> B -> C@^1.0.1
  -> (devDependency) D -> C@1.3.0
```

Yarn will try to install A, B, C, D flat and resolve versions ^1.0.1 and 1.3.0 to a single version.
If production build ignored D branch it might have fetched a different version of C.
Before Yarn people ended up having different non development dependencies (A, B, C) in dev and production mode.

That is why Yarn requires for all the tree to be present at resolution time.
There are ways around this that would satisfy your case without making changes to Yarn, e.g. use a `file:` or `link:` specifier for the local dependency in dev mode and provide a fake one in your production environment.

**Comment by Zephir77167** [comment]
Thanks for the explanation!
Wouldn't resolving devDependencies last (quite a naive approach) be a solution? This way, no dependencies would depend on devDependencies.

**Comment by bestander** [comment]
The order does not matter much, Yarn still tries to de-duplicate deep dependencies (C in the example above) and move them to the root of node_modules.

**Comment by Zephir77167** [comment]
I read [this post](https://hackernoon.com/its-ok-to-not-use-yarn-f28dc766ef32) a few weeks ago, which states that "npm 5 has stronger guarantees across versions and has a stronger deterministic lockfile". Then how can it be than it works without resolving devDependencies? Core implementation? (I'm trying to fully understand why the problem can't be "fixed" here. I understand there are workarounds which can be used to make Yarn do the job)

**Comment by bestander** [comment]
@Zephir77167, this is a great discussion to have but this issue is not a place to have it.
Feel free to jump into the chat https://discordapp.com/invite/yarnpkg and discuss it with the community.

**Comment by Zephir77167** [comment]
Alright, thanks a lot for taking the time to answer!

**Comment by ahmadnassri** [comment]
still seeing this with v1.3.2

**Comment by heisian** [comment]
OK, but why include `devDependencies` in the `--production` dependency tree to begin with?? Isn't the whole point not to include/resolve ANY of those `devDependencies` and their dependencies? I don't care about hoisting `devDependencies`' dependencies.

**Comment by rally25rs** [comment]
@heisian Yeah, your statement is basically correct. It's to produce a deterministic node_modules hoisting tree, which is one of Yarn's key features. I wasn't part of the original team that created Yarn and implemented all that, so I'm not sure if having to re-resolve devDeps was an actual design decision, or just a side-effect of the implementation. There has been talk of putting the hoisting tree location in `yarn.lock` which would mean Yarn would no longer have to resolve those devDeps, but that would change the structure of the lock file, so would likely warrant a major-version change (yarn v2) and hasn't been worked on yet as far as I know.

**Comment by bestander** [comment]
@rally25rs gave a good answer.
I'll just add a few notes.

1. The reason why devDependencies are important in `--production` install is so that you have the same set of dependencies both in your dev environment and production environment.
For example, you can have `left-pad@1.0.1` in devDependencies and `left-pad@^1.0.0` somewhere deep in sub-dependencies. 
Yarn  would resolve  it to the same version `1.0.1` both for dev and production builds and you won't have surprises with an unintended version bump.

2. Calcifying hoisting tree in yarn.lock seems to have quite a few disadvantages.
But we were talking about trying this as an experimental opt-in feature, feel free to fork and propose it.

**Comment by heisian** [comment]
OK, thank you both for the additional clarification.

My main use case for hoisting tree in yarn.lock would be to reduce Docker build times and image size. A couple of the Webpack plugins we use compile binaries at install, leading to long builds times if devDependencies are resolved.

The root cause is that our clients and servers share the same repo, and we use `devDependencies` as a dumping ground for all client-related build dependencies. If we opted to split their `package.json`s up this would be a non-issue.

thanks again.

**Comment by bestander** [comment]
Got it.

So Yarn should only resolve (download) the dev dependencies but it should not be linking (running install scripts) them. Is it what you observe?

Technically Yarn should not need to download the dev dependencies during `--production` install if you already have an up-to-date yarn.lock file.
Quite possible that we have a corner cut here and Yarn fetches and unzips those dependencies even though they are not copied to node_modules at the final install.
If that happens then it is a great opportunity to improve Yarn here, please go ahead and send a PR.

Here are a few ideas that might help your situation:

1. Have you tried offline mirror feature and committing the .tgz files?
Even if you have 1000 dependencies total, it should be around 30MB on disk, maybe that would make build times faster.

2. What if you production app is a workspace and you dev dependencies is a workspace root?
Then you could deploy to docker only the prod workspace.

3. You could strip devDependencies from package.json before sending your app for prod build.
Yarn.lock will still be used but the missing deps will be ignored.

**Comment by heisian** [comment]
I agree with (2), could be a clean and definitive way to go an help with our own file structure.

(3) is also a good-sounding (and very easy) option.

I've not tried the offline mirror feature before, perhaps I will take a look at that.

I think I am observing the install scripts being run, b/c `node-sass` will go through compilation with `node-gyp`, and my resulting Docker image size balloons to ~180MB, whereas if I strip the `devDependencies` altogether, there is no compilation and the image size is around 68MB.

I will see if I can start down the path of finding out what is causing this and make a PR if I do. thanks for the suggestions! :]

**Comment by xPaw** [comment]
This is still an issue, and can even stop the installation when a dev dependency fails to resolve (e.g. `flatmap-stream`).

Repro:
```
git clone https://github.com/thelounge/thelounge
git checkout v3.0.0-rc.5
yarn --production
```

Which ends up with:
```
yarn install v1.12.3
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
error https://registry.yarnpkg.com/flatmap-stream/-/flatmap-stream-0.1.1.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "Unexpected end of data"
```

`flatmap-stream` along with `event-stream` are only required in a devDependency, so there is absolutely no reason as to why Yarn should try to resolve them.

**Comment by mkopinsky** [comment]
I ran into this in a similar scenario. My build server has an old version of node which isn't compatible with eslint 5 (which is a devDependency), but `yarn install --production` still fails with `error eslint@5.13.0: The engine "node" is incompatible with this module. Expected version "^6.14.0 || ^8.10.0 || >=9.10.0".`

**Comment by olingern** [comment]
@mkopinsky This doesn't have anything to do with `production`. You need to give your `package.json` a range for valid node versions.

https://yarnpkg.com/lang/en/docs/package-json/#toc-engines

From the docs:
```
{
  "engines": {
    "node": ">=4.4.7 <7.0.0"
  }
}
```

**Comment by mkopinsky** [comment]
I'm not sure we're understanding each other.

eslint's engine constraint is totally fine. My issue is that I wouldn't expect the engine constraint of a dev dependency to block a production install.

**Comment by arcanis** [comment]
I think this is purely subjective and doesn't take the overall architecture in mind. I'm tempted to close as wontfix.

**Comment by arcanis** [comment]
Giving it a second thought the problem you mention is a bug, @mkopinsky, but it's a different topic as what is discussed in this thread. You should open a new issue.

**Comment by akshetpandey** [comment]
I ran into an issue related to this, since yarn tries to resolve devDependencies when installing in production it fails when the devDependency is not resolvable. This happens if the devDeps are part of a unpublished mono-repo but aren't needed anymore after the compilation is done and a package published.

I second a vote for embedding this information in the yarn.lock file so devDeps are required to be resolvable.

**Comment by mawburn** [comment]
Not sure if this will help anyone or not, but I keep running into this issue with some quality of life tools I use for development that aren't available on our CI/CD platform. 

I keep running into the error:

>The platform "linux" is incompatible with this module.

The offender is only defined in my `devDependencies`. To work around it, I added the `--ignore-platform` tag.

    yarn install --production --ignore-platform

This is obviously a hacky work around that could get you in trouble if you're not careful. I really wish yarn wouldn't try to resolve packages I don't want it to install.

**Comment by billouboq** [comment]
This is a really big problem when deploying to Heroku because bundle size is capped at 500mo max and some devDependencies are really big.. Can we get rid of the devDependencies installation ?

**Comment by BYK** [comment]
>This is a really big problem when deploying to Heroku because bundle size is capped at 500mo max and some devDependencies are really big.. Can we get rid of the devDependencies installation ?

The issue is not installing `devDependencies` in production mode, it is merely resolving them. If your Heroku instance is trying to install `devDependencies`, make sure your `NODE_ENV` is set to `production`.

**Comment by Akryum** [comment]
@BYK `--prefer-offline --frozen-lockfile` doesn't work :(

**Comment by armordog** [comment]
@BYK By "it is merely resolving them" do you mean it is "adding them to node_modules"?
Because for most packages that indistinguishable from installing them.

If you _don't_ mean that, then this flag doesn't work at all (v1.21.1)

**Comment by imagine10255** [comment]
Very puzzled, the packages appearing in dependencies and devDependencies should be unique.
How can there be the same package?

**Comment by BYK** [comment]
>@BYK By "it is merely resolving them" do you mean it is "adding them to node_modules"?
>Because for most packages that indistinguishable from installing them.

Resolving means reading their `package.json` to determine the whole install tree. It does _not_ put them under `node_modules`. That step is the linking step.

>If you don't mean that, then this flag doesn't work at all (v1.21.1)

It does since we have tests for this behavior. It maybe getting overridden by your environment variables or something else.

**Comment by dvdotsenko** [comment]
Use of `--frozen-lockfile` (as in `yarn install --prod --frozen-lockfile` ) AVOIDS pulling (downloading, "installing") dev dependencies.  You just need to resolve `yarn.lock` elsewhere before.

Tested with yarn 1.17.3

**Comment by IlyaSemenov** [comment]
@dvdotsenko At least not if a dev dependency points to an external Git repo... Yarn wants to fetch the remote repo when I run `yarn install --production --frozen-lockfile` in a CI/CD pipeline.

**Comment by mattnathan** [comment]
I know this issue is old, but as this isn't resolved and my specific case isn't mentioned I thought I'd chip in.

The project I work on is a js monorepo using Firebase (Firestore, functions, hosting, etc). We've defined `.d.ts` definitions for the data we store in Firestore and keep those in a `types` directory in the project root. The functions located in `functions/` reference these types as `devDependencies` using `file:../types`.

During deploy `yarn install --production` is run on the firebase server using the functions `package.json` - this is completely out of our control to configure. This doesn't work because the `file:../types` fails to resolve even though it's not needed for production, only the `functions/` directory is actually uploaded.

Obviously we could use one of the work arounds mentioned here (copy/link/embed the files, use NPM5, etc) but they are all work arounds involving extra work and are less than ideal.

**Comment by kivi** [comment]
This is not fixed for yarn `1.22.5`

I am using `node:14-buster-slim` in my Dockerfile
`yarn install --production --frozen-lockfile`  also installs devDependencies and that is resulting a huge image.

For now I have a workaround, thanks to @NotWearingPants https://github.com/yarnpkg/yarn/issues/3630#issuecomment-641959126

**Comment by avshyz-gs** [comment]
This is also not fixed in 1.22.17.
An certain eslint plugin requires a different node version from our production image's node version.
I'd follow NotWearingPants's comment.

**Comment by felsspat** [comment]
Really, no solution after 6 years? This is a real problem :(

**Comment by K-Mistele** [comment]
I'm running into this too, running `yarn install --production --frozen-lockfile` still results in a docker image that contains all of my dev dependencies in my `node_modules` directory, resulting in the image being 10x larger than with npm
